### PR TITLE
Add JS3 Week 2 prep content MOVED TO SPRINT 3

### PR DESCRIPTION
**Pull request by Dedekind561**
## What does this change?

Module:
Week(s):

## Checklist

- [ ] I have read the [contributing guidelines](CONTRIBUTING.MD)
- [ ] I have checked my spelling and grammar with an [automated tool](https://www.grammarly.com/grammar-check)
- [ ] I have previewed my changes to check the [markdown renders](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax) as I intend
- [ ] I have run my code to check it works
- [ ] My changes follow our [Style Guide](https://curriculum.codeyourfuture.io/guides/code-style-guide)

### CODE REVIEW

File: content/js3/blocks/fetching-data/index.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+time = 20
+facilitation = false
+emoji= '🧩'
+[objectives]
+    1='Update a problem statement using web APIs to fetch data'
++++
+
+So far we have displayed films data stored in our JavaScript code. But real applications fetch data from servers over the internet. We can restate our problem as follows:
+
+> _Given_ a **web API that serves** film data
```
Comment (illicitonion) [inline content/js3/blocks/fetching-data/index.md:13]: It seems like we're using "web API" to mean both "JS APIs browsers tend to implement" (as per https://developer.mozilla.org/en-US/docs/Web/API) and "APIs served over the web by other servers".

I'm not sure this is useful - should we maybe use another term to mean the latter?

File: content/js3/blocks/asynchrony/index.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+```js
+const search = document.getElementById("search");
+const waitForEvent = search.addEventListener("input", handleInput);
+```
+
+The callback `handleInput` cannot run until the user types. With `fetch`, the callback function cannot run until the data arrives. In both cases, we are waiting for something to happen before we can run our code.
+{{</tab>}}
+{{<tab name="Visualise the Event Loop">}}
+
+<iframe src="http://latentflip.com/loupe/?code=JC5vbignYnV0dG9uJywgJ2NsaWNrJywgZnVuY3Rpb24gb25DbGljaygpIHsKICAgIGNvbnNvbGUubG9nKCdZb3UgY2xpY2tlZCB0aGUgYnV0dG9uIScpOyAgICAKfSk7Cgpjb25zb2xlLmxvZygiSGkhIik7Cgpjb25zb2xlLmxvZygiV2VsY29tZSB0byB0aGUgZXZlbnQgbG9vcCIpOw%3D%3D!!!PGJ1dHRvbj5DbGljayBtZSE8L2J1dHRvbj4%3D" width="100%" height="480px"></iframe>
```
Comment (illicitonion) [inline content/js3/blocks/asynchrony/index.md:58]: Are we expecting the trainees to watch the 27 minute video in the loupe help text?

I haven't watched it yet, but assuming it's good, I think we probably should (and should tell them to). If not, I think we need to explain "call stack" and "callback queue" a bit before people can find this super enlightening?

Either way, I think this block is more than 30 minutes.

File: content/js3/blocks/using-fetch/index.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
++++
+title = '🌐 Requesting from a server side API'
+headless = true
+time = 20
+facilitation = false
+emoji= '🧩'
+[objectives]
+1="List 4 preceding concepts of asynchronous programming in JavaScript"
```
Comment (illicitonion) [inline content/js3/blocks/using-fetch/index.md:8]: Which concepts are these? I'm not sure what they are from the current contents of the prep page...

File: content/js3/blocks/using-fetch/index.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+3="Fetch data from a server side API"
++++
+
+So now we have these pieces of our giant concept map
+
+🗓️ we know that sending 📤 requests over a network takes time
+🧵 we know that we should not stop our program to wait for data
+📤 we know that we can send a request using `fetch()`
+🐕 we know that `fetch` is a 💻 client side 🧰 Web API
+🫱🏿‍🫲🏽 we know that `fetch` returns a `Promise`? (...this last one is pending)
```
Comment (illicitonion) [inline content/js3/blocks/using-fetch/index.md:19]: I suspect this is more confusing than it is funny, to our trainees? Maybe remove?

File: content/js3/blocks/promises/index.md [hunk 5]
```diff
@@ -0,0 +1,9 @@
++++
+title = 'What is a Promise?'
```
Comment (illicitonion) [inline content/js3/blocks/promises/index.md:2]: Is this block meant to have content? I think content would be useful here, but if not, let's remove it?

File: content/js3/blocks/then/index.md [hunk 6]
```diff
@@ -0,0 +1,9 @@
++++
+title = 'Making sense of .then()'
```
Comment (illicitonion) [inline content/js3/blocks/then/index.md:2]: Ditto about missing content

File: content/js3/blocks/chaining/index.md [hunk 7]
```diff
@@ -0,0 +1,8 @@
++++
+title = 'Chaining Promises'
```
Comment (illicitonion) [inline content/js3/blocks/chaining/index.md:2]: Missing content? Remove the block until it's used?

**Review by illicitonion — COMMENTED** [review]
Generally looks good - I opened up https://github.com/CodeYourFuture/curriculum/pull/462 with some copy edits.

I'm not sure I'm sold on the "using promises without explaining promises at all" part... I'd maybe be tempted to remove the `.then` and "`console.log` a promise and look at it" piece, and just give the replacement code and say "We'll learn more about how it works next sprint"... WDYT?

File: content/js3/blocks/chaining/index.md (hunk 7 above)
Comment (SallyMcGrath) [reply content/js3/blocks/chaining/index.md:2]: FYI Daniel I think there's a bunch of stuff to go on this, but I had to move to other urgent work. Will look when I get a chance

File: content/js3/blocks/then/index.md (hunk 6 above)
Comment (SallyMcGrath) [reply content/js3/blocks/then/index.md:2]: oh yes I just set up these blocks based on the headings -- then we reordered so these will all be for week 4

File: content/js3/blocks/promises/index.md (hunk 5 above)
Comment (SallyMcGrath) [reply content/js3/blocks/promises/index.md:2]: yes it's for week 4 now -- we reordered

File: content/js3/blocks/using-fetch/index.md (hunk 4 above)
Comment (SallyMcGrath) [reply content/js3/blocks/using-fetch/index.md:19]: the whole thing? or just the last line

It's good to recap what we know. I will remove jokes no jokes allowed I know

🧵 we know that we should not stop our program to wait for data
📤 we know that we can send a request using `fetch()`
🐕 we know that `fetch` is a 💻 client side 🧰 Web API
🫱🏿‍🫲🏽 we don't yet know that `fetch` returns a `Promise`, and we will investigate this next time

File: content/js3/blocks/asynchrony/index.md (hunk 2 above)
Comment (SallyMcGrath) [reply content/js3/blocks/asynchrony/index.md:58]: Yes please update the time!

I like the video a lot and have used it with trainees. I will say that about half of them liked it and half of them preferred just using the loupe, and only really got the video after playing with the visualiser. This is pretty classic -- some people are principles first learners and some are practical first.

File: content/js3/blocks/fetching-data/index.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
++++
+title = '🐕 Fetching data'
+headless = true
+time = 20
+facilitation = false
+emoji= '🧩'
+[objectives]
+    1='Update a problem statement using web APIs to fetch data'
```
[the remark this answers is not in the document]
Comment (SallyMcGrath) [reply content/js3/blocks/fetching-data/index.md:8]: Please offer line changes where you can as I am rewriting 3 modules in 1 week 🙏 🙏

File: content/js3/blocks/fetching-data/index.md (hunk 1 above)
Comment (SallyMcGrath) [reply content/js3/blocks/fetching-data/index.md:13]: Ah yes I think this is a mismatch between me and Mitch.

**Review by illicitonion — APPROVED** [review]
LGTM except both the mermaid and the loupe embeds seem broken?

Screenshot 2023-12-26 at 22 39 33

Screenshot 2023-12-26 at 22 40 01

**Pull request merged**